File: apps/api/src/infra/imagekit-url-transformer.ts
const IMAGEKIT_DOMAIN = "ik.imagekit.io";
const DEFAULT_AVATAR_SIZE = 256;

export interface TransformOptions {
  width?: number;
  height?: number;
}

export function transformImageKitUrl(
  url: string | null | undefined,
  options: TransformOptions = {},
): string | null {
  if (!url) return null;
  if (!isImageKitUrl(url)) return url;

  const width = options.width ?? DEFAULT_AVATAR_SIZE;
  const height = options.height ?? DEFAULT_AVATAR_SIZE;

  const transformations = `tr=w-${width},h-${height},q-100,f-auto`;

  const uri = new URL(url);
  uri.search = transformations;

  return uri.toString();
}

function isImageKitUrl(url: string): boolean {
  try {
    const uri = new URL(url);
    return uri.host.includes(IMAGEKIT_DOMAIN);
  } catch {
    return false;
  }
}

File: apps/mobile/lib/core/utils/imagekit_url_transformer.dart
class ImageKitUrlTransformer {
  static const String _imageKitDomain = 'ik.imagekit.io';
  static const int _defaultSize = 256;

  static String? transformUrl(String? url, {int? size}) {
    if (url == null) return null;
    if (url.isEmpty) return '';
    if (!isImageKitUrl(url)) return url;

    final targetSize = size ?? _defaultSize;
    final transformations = 'tr=w-$targetSize,h-$targetSize,q-100,f-auto';

    final uri = Uri.parse(url);

    return Uri(
      scheme: uri.scheme,
      host: uri.host,
      path: uri.path,
      query: transformations,
    ).toString();
  }

  static bool isImageKitUrl(String? url) {
    if (url == null || url.isEmpty) return false;
    try {
      final uri = Uri.parse(url);
      return uri.host.contains(_imageKitDomain);
    } catch (_) {
      return false;
    }
  }
}
